libs/stats/odc/stats/tasks.py

```diff
@@ -116,7 +178,34 @@ def save(
         :param predicate: If supplied filter Datasets as they come in with custom filter, Dataset->Bool
         :param msg: Observe messages if needed via callback
         :param debug: Dump some intermediate state to files for debugging
+        :param dss: A generator of datasets to use
+        :param n_dss: The number of datasets in the generator
         """
+        
+        # either product and dc must be set, or dss and n_dss must be
+        valid_args = (product is not None) and (dc is not None)
+        valid_args |= (dss is not None) and (n_dss is not None)
```

Comment:
I'd say `n_dss` needs to be optional, we won't know how many virtual datasets will come out, we can only know upper bound really, which is `min(count(p) for p in products)`

Reply:
I think we still need an upped bound for `tqdm` to work right?

Reply:
it supports unbounded as well, won't be as useful, but will work